src/libos/src/vm/vm_epc.rs

```diff
@@ -384,22 +357,209 @@ pub fn enclave_page_fault_handler(
 ) -> Result<()> {
     let pf_addr = exception_info.faulting_address as usize;
     let pf_errcd = exception_info.error_code;
-    trace!(
+    debug!(
         "enclave page fault caught, pf_addr = 0x{:x}, error code = {:?}",
-        pf_addr,
-        pf_errcd
+        pf_addr, pf_errcd
     );
 
     USER_SPACE_VM_MANAGER.handle_page_fault(rip, pf_addr, pf_errcd, kernel_triggers)?;
 
     Ok(())
 }
 
-extern "C" {
-    fn occlum_ocall_mprotect(
-        retval: *mut i32,
-        addr: *const c_void,
-        len: usize,
-        prot: i32,
-    ) -> sgx_status_t;
+mod edmm_api {
+    use super::*;
+    use std::marker::PhantomData;
+    use std::mem;
+
+    pub(super) struct EDMMLocalApi;
+
+    impl EDMMLocalApi {
+        // To replace sgx_mm_commit
+        pub(super) fn commit_memory(start_addr: usize, size: usize) -> Result<()> {
+            let si = sec_info_t::new_for_commit_memory();
+            for page in (start_addr..start_addr + size).step_by(PAGE_SIZE) {
+                let ret = unsafe { do_eaccept(&si as *const sec_info_t, page) };
+                if ret != 0 {
+                    return_errno!(EFAULT, "do_eaccept failure");
+                }
+            }
+            Ok(())
+        }
+
+        // To replace sgx_mm_commit_data
+        pub(super) fn commit_with_data(addr: usize, data: &[u8], perm: VMPerms) -> Result<()> {
+            let si = sec_info_t::new_for_commit_with_data(&perm);
+            let size = data.len();
+            let mut src_raw_ptr = data.as_ptr() as usize;
+            for dest_page in (addr..addr + size).step_by(PAGE_SIZE) {
+                let ret =
+                    unsafe { do_eacceptcopy(&si as *const sec_info_t, dest_page, src_raw_ptr) };
+                if ret != 0 {
+                    return_errno!(EFAULT, "do_eacceptcopy failure");
+                }
+
+                Self::modify_permissions(dest_page, PAGE_SIZE, VMPerms::DEFAULT, perm)?;
```

Comment:
`Self::modify_permissions(....);` move to line 405;

Reply:
Done